pkg/pillar/cmd/zedagent/handleProfile.go

```diff
@@ -0,0 +1,326 @@
+// Copyright (c) 2021 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package zedagent
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	zconfig "github.com/lf-edge/eve/api/go/config"
+	"github.com/lf-edge/eve/api/go/profile"
+	"github.com/lf-edge/eve/pkg/pillar/flextimer"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/zedcloud"
+)
+
+const (
+	defaultLocalProfileServerPort = "8888"
+	savedLocalProfileFile         = "lastlocalprofile"
+)
+
+// urlAndSrcIP structure for mapping source IP and url of local server
+type urlAndSrcIP struct {
+	srcIP     net.IP
+	actualURL string
+}
+
+func getLocalProfileURL(localProfileServer string) (string, error) {
+	localProfileURL := fmt.Sprintf("http://%s", localProfileServer)
+	u, err := url.Parse(localProfileURL)
+	if err != nil {
+		return "", fmt.Errorf("url.Parse: %s", err)
+	}
+	if u.Port() == "" {
+		localProfileURL = fmt.Sprintf("%s:%s", localProfileURL, defaultLocalProfileServerPort)
+	}
+	return fmt.Sprintf("%s/api/v1/local_profile", localProfileURL), nil
+}
+
+//checkAndHandleLocalProfileUpdate gets local profile config if localProfileServer defined,
```

Comment:
Actually this gets called when one of LocalProfile, GlobalProfile, or LocalProfileServer changes.
Thus the name is misleading; perhaps naming it checkAndHandleProfileUpdate()?

Reply:
It is only about localProfile we try to get from provided url. The main purpose is to not block main loop with requests. And the trigger tell about changes in url.
Actually, I cannot catch the problem with spreading processing onto two places with different sources: in this place I obtain and process localProfile comes from localProfileServer and in the second place (parseProfile) I obtain and process parameters coming from EdgeDevConfig (or depends on it).
Probably I can move to some function block with determine, checking for change of currentProfile and publishing in case of change (it is duplicated here and in parseProfile).

Reply:
The problem is making the code the maintainable. Any code can be made to work.
Since this is subtle it is critical to have the state machine in one place and not distributed across different functiions in very different places.

Reply:
I have a suggestion.
A single function (this function renamed to profileStateMachine or handleProfileChanges) which has a boolean argument "skipFetch". That function does all of the state changes. Call it with skipFetch true from the parser, and skipFetch false from the task.